WebP animations

MaartenBent opened the pull request:
This extends #24479 by addressing some of the remarks made there, adding libwebp as a built-in submodule, and adding WebP animation support.
Also see #23489

https://github.com/user-attachments/assets/c5190d7b-ead8-4168-8d6b-1b91c82b91c4

I think it is mostly finished. Except for updating all the build files (only CMake is up-to-date currently). And documentation. And addressing some more remarks from the linked PR. And maybe adding support for saving WebP animations.

hoehermann committed:
Add support for WebP image file format via libwebp.

MaartenBent committed:
Use WebP as built-in submodule

MaartenBent committed:
Update wxWEBPHandler

Add function to get the WebP library version.

Add LoadAnimation function to get all frame images, background colour and duration.

Use WebPAnimDecoder to decode frames other than the default (-1) frame.
This ensures all returned frames will have the same image size.
WebPDemuxGetFrame can return frames smaller than the full image size, but there is no way to get the offset of the frame.

Reduce code duplication when creating rgba image.

Add enum to store image format (undefined/mixed, lossy or lossless).
Add lossless and lossy quality options to wxImage for saving WebP.

Add wxBITMAP_TYPE_WEBP_RESOURCE because all other image handlers have this too.

MaartenBent committed:
Add WebP image to image sample

MaartenBent committed:
Add wxWebPDecoder for WebP animations

MaartenBent committed:
Show frames of animated images in image sample

MaartenBent committed:
Add WebP format to animation sample

Improve high-dpi support.

PBfordev commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,346 @@
+    {
+        // Default first frame, Use RGB(A) decoder
+        // We can't use this for other frames, because the frame might be sub-frames, smaller than the
+        // full image. It's size is known, but the x_offset and y_offset can not be queried, so we can't
```
```suggestion
        // full image. Its size is known, but the x_offset and y_offset can not be queried, so we can't
```

PBfordev commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,346 @@
+bool wxWEBPHandler::DoCanRead(wxInputStream& stream)
+{
+    // check header according to https://developers.google.com/speed/webp/docs/riff_container
+    const std::string riff = "RIFF";
```
Would it not be better to use `constexpr char*` instead of using and creating four `std::string` here?

PBfordev commented on tests/test.bkl:
```diff
@@ -364,7 +365,8 @@
         <files>horse.ani  horse.bmp  horse.cur  horse.gif  horse.ico  horse.jpg
-               horse.pcx  horse.png  horse.pnm  horse.tga  horse.tif  horse.xpm</files>
+               horse.pcx  horse.png  horse.pnm  horse.tga  horse.tif  horse.xpm
+               horse.webp</files>
```
I know you said you did not modify all the build files except for CMake, but perhaps you could still add `horse.webp` also to `samples/image/image.bkl`...

PBfordev reviewed (commented):
Thanks Maarten for the thankless tasks of finishing the PR.

Aside to specific nitpicks, it seems there are no checks for failed memory allocations (nor errors when reading from streams)? I am not sure what is wxWidgets policy for checking out-of-memory errors, in places with large allocations in particular...

vadz commented on configure.ac:
```diff
@@ -7719,6 +7752,10 @@ if test "$wxUSE_LIBLZMA" = "yes"; then
     fi
 fi
+if test "$wxUSE_LIBWEBP" = "yes"; then
+    dnl no need to check for duplicates since webp is disabled in the our builtin tiff
```
```suggestion
    dnl no need to check for duplicates since webp is disabled in our builtin tiff
```

vadz commented on include/wx/imagwebp.h:
```diff
@@ -0,0 +1,62 @@
+#if wxUSE_LIBWEBP
+
+#define wxIMAGE_OPTION_WEBP_QUALITY  wxT("WebPQuality")
+#define wxIMAGE_OPTION_WEBP_LOSSLESS wxT("WebPLossless") // 0 = undefined (/mixed), 1 = lossy, 2 = lossless
```
We probably should have a (scoped) enum for these values?

vadz commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,346 @@
+    #include "wx/log.h"
+#endif
+
+wxWebPAnimationFrame::wxWebPAnimationFrame()
+    : image(wxNullImage)
+    , bgColour(wxNullColour)
+    , duration(0)
+{
+}
+
```
As mentioned in the comment in the header:

```suggestion
```

vadz commented on include/wx/imagwebp.h:
```diff
@@ -0,0 +1,62 @@
+#define wxIMAGE_OPTION_WEBP_LOSSLESS wxT("WebPLossless") // 0 = undefined (/mixed), 1 = lossy, 2 = lossless
+
+
+struct WXDLLIMPEXP_CORE wxWebPAnimationFrame
+{
+    wxImage image;
+    wxColour bgColour;
+    int duration;
+
+    wxWebPAnimationFrame();
+};
```
I'd rather get rid of the trivial ctor and then avoid exporting this struct from the DLL too:

```suggestion
struct wxWebPAnimationFrame
{
    wxImage image;
    wxColour bgColour;
    int duration = 0;
};
```

vadz commented on include/wx/webpdecoder.h:
```diff
@@ -0,0 +1,57 @@
+    virtual wxColour GetTransparentColour(unsigned int frame) const override;
+
+    virtual bool Load(wxInputStream& stream) override;
+    bool ConvertToImage(unsigned int frame, wxImage* image) const override;
+
+    wxAnimationDecoder* Clone() const override
+    {
+        return new wxWebPDecoder;
+    }
+    wxAnimationType GetType() const override
+    {
+        return wxANIMATION_TYPE_WEBP;
+    }
```
Just for consistency with the functions above:

```suggestion
    virtual bool ConvertToImage(unsigned int frame, wxImage* image) const override;

    virtual wxAnimationDecoder* Clone() const override
    {
        return new wxWebPDecoder;
    }
    virtual wxAnimationType GetType() const override
    {
        return wxANIMATION_TYPE_WEBP;
    }
```

vadz commented on src/common/webpdecoder.cpp:
```diff
@@ -0,0 +1,85 @@
+/////////////////////////////////////////////////////////////////////////////
+// Name:        src/common/webpdecoder.cpp
+// Purpose:     wxWebPDecoder, WebP Animation decoder
+// Author:
+// Copyright:
```
A bit weird to leave these lines empty.

vadz commented on interface/wx/gdicmn.h:
```diff
@@ -46,6 +46,7 @@ enum wxBitmapType
     wxBITMAP_TYPE_TGA,
     wxBITMAP_TYPE_MACCURSOR,
     wxBITMAP_TYPE_MACCURSOR_RESOURCE,
+    wxBITMAP_TYPE_WEBP,
```
```suggestion
    wxBITMAP_TYPE_WEBP, ///< @since 3.3.0
```

vadz commented on interface/wx/image.h:
```diff
@@ -660,6 +663,7 @@ class wxImage : public wxObject
             @li wxBITMAP_TYPE_ICO: Load a Windows icon file (ICO).
             @li wxBITMAP_TYPE_CUR: Load a Windows cursor file (CUR).
             @li wxBITMAP_TYPE_ANI: Load a Windows animated cursor file (ANI).
+            @li wxBITMAP_TYPE_WEBP: Load a WebP file.
```
I don't know if it should be repeated everywhere, but we probably do need to mention that WebP support is only available since 3.3.0.

vadz commented on misc/scripts/check_mixed_eol.sh:
```diff
@@ -16,7 +16,7 @@ while IFS= read -r -d '' file; do
     # skip binary files
     case "$file" in
-    *.ani | *.bmp | *.chm | *.cur | *.dia | *.gif | *.gz | *.hlp | *.icns | *.ico | *.jpg | *.mo | *.mpg | *.pcx | *.pdf | *.png | *.pnm | *.pyc | *.tga | *.tif | *.ttf | *.wav | *.zip )
+    *.ani | *.bmp | *.chm | *.cur | *.dia | *.gif | *.gz | *.hlp | *.icns | *.ico | *.jpg | *.mo | *.mpg | *.pcx | *.pdf | *.png | *.pnm | *.pyc | *.tga | *.tif | *.ttf | *.webp | *.wav | *.zip )
```
Extensions are supposed to be in alphabetical order here:
```suggestion
    *.ani | *.bmp | *.chm | *.cur | *.dia | *.gif | *.gz | *.hlp | *.icns | *.ico | *.jpg | *.mo | *.mpg | *.pcx | *.pdf | *.png | *.pnm | *.pyc | *.tga | *.tif | *.ttf | *.wav | *.webp | *.zip )
```

vadz commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,346 @@
+#if wxUSE_IMAGE && wxUSE_LIBWEBP
+
+#include "wx/imagwebp.h"
+#include "webp/demux.h"
+#include "webp/decode.h"
+#include "webp/encode.h"
```
Those should use `<...>` include style.

And I'd also add blank lines before and after them as they're a different group of headers.

vadz commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,346 @@
+#include "webp/encode.h"
+#include <vector>
+
+#ifndef WX_PRECOMP
```
Very minor, but I'd put this before all the other headers for consistency with the other files.

vadz commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,346 @@
+#include <functional>
+
+namespace
+{
```
I'd like to unindent the contents of this namespace for consistency.

vadz reviewed (commented):
Thanks a lot for working on this!

I don't see any big problems, but there are, as you can see, quite a few small remarks (some of which I had already left in the original PR, but completely forgot about them...) — and I also agree with the comments already left by @PBfordev. I think I can count on you dealing with them when you have time but if you'd like any help with it (or disagree with any of them), please let me know.

FWIW I don't think it's critical to have this in 3.3.0, as it's a pure addition, we could add it in 3.3.1 too, but it would be nice to have it there, of course.

Thanks again!

MaartenBent commented:
Thanks @PBfordev and @vadz for the reviews. I'll address all the issues. But if 3.3.0-RC is planned for this month it'll probably not make it.

The only thing I would probably need help with (if we want this), is adding support to build the built-in version with configure. Similar to how it builds libtiff.

MaartenBent committed:
Update and generate build files

Run upmake, bakefile_gen, makeprojects.py, update-setup-h, autogen.

MaartenBent committed:
CMake: Fix iOS OpenGL warning

WebP sub-project searches for OpenGL, and sets the OpenGL variables to NOT_FOUND.
Make sure all OpenGL variables are updated when OpenGL dependency is set manually.

MaartenBent reviewed (commented):
I think I addressed all of the review comments.

Questions:
- Should I rename `wxWEBPHandler` to `wxWebPHandler`? All image handlers have capitalized names, but all WebP classes don't.
- Where should the submodule go? `src/webp` or `3rdparty/webp`?
- I added `wxBITMAP_TYPE_WEBP_RESOURCE` because all formats have this. Is this fine?

Some things that I won't be adding (unless absolutely required):
- Saving animated webp
- Using built-in webp from configure (or msvs projects)

vadz commented:
> I think I addressed all of the review comments.

Thanks a lot!
 
> Questions:
> 
>     * Should I rename `wxWEBPHandler` to `wxWebPHandler`? All image handlers have capitalized names, but all WebP classes don't.

I guess we should keep this name for consistency, even although I find `wxWebPHandler` more readable too.

>     * Where should the submodule go? `src/webp` or `3rdparty/webp`?

I don't have strong preferences here, but I prefer `3rdparty`.
 
>     * I added `wxBITMAP_TYPE_WEBP_RESOURCE` because all formats have this. Is this fine?

Hmm, this is not going to be implemented anywhere is it?

vadz commented on include/wx/imagwebp.h:
```diff
@@ -0,0 +1,72 @@
+class WXDLLIMPEXP_CORE wxWEBPHandler : public wxImageHandler
+{
+public:
+    inline wxWEBPHandler()
```
Very minor, but we typically don't use `inline` expicitly

```suggestion
    wxWEBPHandler()
```

vadz commented on interface/wx/image.h:
```diff
@@ -1434,6 +1441,16 @@ class wxImage : public wxObject
+            (best quality, larger output).
+            When unspecified, quality 90 will be used.
+        @li @c wxIMAGE_OPTION_WEBP_LOSSLESS: Format type of the image that is read, matching
+            the WebPBitstreamFeatures format with: 0 = undefined (/mixed), 1 = lossy, 2 = lossless
```
Shouldn't we mention the symbolic names for these constants here?

vadz commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,365 @@
+typedef std::unique_ptr<WebPAnimDecoder, std::function<void(WebPAnimDecoder*)>> WebPAnimDecoderPtr;
+typedef std::unique_ptr<uint8_t, std::function<void(uint8_t*)>> WebPDecodeRGBAPtr;
+
+void rgbaToImage(wxImage* image, uint8_t* rgba)
```
FWIW I think this could be moved to `wxImage` itself.

vadz commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,365 @@
+    }
+}
+
+bool copyImageData(wxImage* to, wxImage& from)
```
And this is already part of `wxImage`, it's called `Copy()` in it, so should be replaced with the call to this function.

vadz commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,365 @@
+        return false;
+    }
+
+    if (!image->Create(features.width, features.height, false)) {
```
Style nitpick:

```suggestion
    if (!image->Create(features.width, features.height, false))
    {
```

vadz commented on src/common/imagwebp.cpp:
```diff
@@ -0,0 +1,365 @@
+    if (useFrameDemuxer)
+    {
+        // Default first frame, Use RGB(A) decoder
+        // We can't use this for other frames, because the frame might be sub-frames, smaller than the
+        // full image. Its size is known, but the x_offset and y_offset can not be queried, so we can't
+        // reconstruct the full-size image.
```
Minor, but would be nice to reformat to fit in 80 columns.

vadz commented on tests/image/imagwebp.cpp:
```diff
@@ -0,0 +1,74 @@
+    CHECK(!webpHandler.LoadFile(&loadingImage, inputStream, false));
+}
+
+static bool WebPRoudtrip(wxImage& savingImage, wxImage& loadingImage)
```
Typo:
```suggestion
static bool WebPRoundtrip(wxImage& savingImage, wxImage& loadingImage)
```

vadz reviewed (commented):
Thanks a lot once again!

This globally looks good to merge to me. Ideal would probably be to build libwebp when using MSVS, just as we do with the other libraries, but this is not a hard requirement, whoever is interested in using it can always implement this later.

MaartenBent committed:
Add wxImage::SetDataRGBA

MaartenBent reviewed (commented):
> I don't have strong preferences here, but I prefer 3rdparty.

I'll use this location then, once the fork is created.

> Hmm, this is not going to be implemented anywhere is it?

No. I just copy-paste all the `wxBITMAP_TYPE_GIF`/`wxBITMAP_TYPE_GIF_RESOURCE` and added the same for WebP.
`wxBITMAP_TYPE_GIF_RESOURCE` also doesn't have any use.

> Ideal would probably be to build libwebp when using MSVS, just as we do with the other libraries, but this is not a hard requirement, whoever is interested in using it can always implement this later

Yes, I can try to add it after this PR. Adding the files shouldn't be too difficult. Usually it us dealing with the generated `config.h` file.

MaartenBent committed:
Update wxWEBPHandler tests

MaartenBent committed:
Update wxWEBPHandler and wxWebPDecoder documentation

MaartenBent committed:
Invalidate wxGenericAnimationCtrl size when fitting image

MaartenBent commented:
I removed the libwebp msvc project, I'll add it to #25339 instead.

I moved the submodule to `3rdparty/libwebp` and cleaned up all the commits. Only remaining change is using the correct submodule remote.
For that, the [wx branch](https://github.com/wxWidgets/libwebp/tree/wx) should point to tag `v1.5.0`, and I'll add a PR with some minimal changes https://github.com/wxWidgets/libwebp/commit/eb311fd27f4fd14ef4a6bdd21866c4c8236919be

vadz commented:
I've reset `wx` branch now (by deleting the old one and recreating it in the right place, hopefully this didn't break anything).

MaartenBent committed:
Make wxImage::SetDataRGBA argument const

MaartenBent commented:
@vadz this and #25339 are now ready.

The last commit ([Make wxImage::SetDataRGBA argument const](https://github.com/wxWidgets/wxWidgets/pull/25205/commits/5dd62901583aee18fa6ecca40c2c3f656f3964f5)) is not in the other PR, so I propose to merge this PR first and then #25339 (instead of only #25339).
